Rework source address determination of wildcard sockets

koflin opened the pull request.

koflin commented:
@marcfrei @tzaeschke I updated the diagram a bit, does this make sense in your opinion?

marcfrei commented:
Looks very good. Should we include the diagram into a "design" sub-directory in "docs" (or similar)? If yes, here some hints on how to streamline the wording a little more:

- "only allowed for servers" -> "only possible for servers"
- "take destination address of first packet received as source address" -> "take underlay destination address of packet received as source address" (for every packet individually)
- "does the network already know the source address?" -> "source address for network provided?"
- "did connect yield 0.0.0.0?" -> "did connect yield source address?"

koflin commented:
> Looks very good. Should we include the diagram into a "design" sub-directory in "docs" (or similar)? If yes, here some hints on how to streamline the wording a little more:

Yes, we can do that. Thank you for the suggestions for the wording.
> 
> * "only allowed for servers" -> "only possible for servers"
> * "take destination address of first packet received as source address" -> "take underlay destination address of packet received as source address" (for every packet individually)

Because of our assumption stated above, it suffices if we take the destination address of the **first** incoming packet as our source address and store it on the network and socket.
> * "does the network already know the source address?" -> "source address for network provided?"

I rephrased this a bit differently. What I mean by this is that as soon as one socket knows its source address it can set this on the network and then it's considered **known** on the network.

> * "did connect yield 0.0.0.0?" -> "did connect yield source address?"

I'm not sure this makes things clearer. I want to emphasis here that even though we connect, we can still receive 0.0.0.0 when getting the name of the connected socket and thus don't know the source address we can use.

koflin commented:
> Yes, looks good. One question though: "connect to arbitrary border router and determine source address of socket" -> can this really yield 0.0.0.0?

Yes it can. It actually happens on ESP32S3 for me. I think the general issue here is that this is implementation defined and we are not guaranteed to get the actual source address after this.

I think this whole process of determining the source address is similar to the SCION end-host boostrapping process. We try to use different methods and hints to get the correct source address, but it may be that all of them fail. Then as a last resort the user has to provide the source address manually.

> 
> Also, maybe rephrase "take destination address of first packet received as source address" -> "...first incoming packet as source ..."?

Thanks, I adjusted this.

koflin commented:
@marcfrei @tzaeschke I've implemented the source address determination as discussed above. I've also added some documentation in `docs/design`.

I have two remaining questions:

1. I removed the functions `scion_network_set_addr()` and  `scion_network_get_addr()`. I am not sure if I should have done this. My main concern was that this might add some confusion for the users and the functions are not really necessary. Any opinions?
2. Accessing and setting the source address of a network is not yet thread-safe. This can lead to issues when multiple sockets share a network concurrently. Should we ignore this for now?

tzaeschke commented:
How about keeping only the `scion_network_get_addr()` but renaming it to `scion_network_get_local_addr()` or similar?

Removing the set-function should help with concurrency.
For concurrency, can we enforce setting the src address during initialization of the `network`?

koflin commented:
> How about keeping only the `scion_network_get_addr()` but renaming it to `scion_network_get_local_addr()` or similar?
> 
> Removing the set-function should help with concurrency. For concurrency, can we enforce setting the src address during initialization of the `network`?

At the moment we can't enforce this, because sometimes the socket will only know it's source address after it has been bound or received the first packet. At that moment it will then want to set the source address on the network, which already has been initialized.

What I could do is that sockets that got their source address from explicitly binding or receiving the first packet won't write back this source address to the network. Hence, the source address would only need to be set (with connecting to the border router) in the initialization of the network. The drawback there would be that, because sometimes this connecting does not yield the source address sometimes a network won't have a source address set after initialization. We then probably also need the `scion_network_set_addr()` again to be able to set the source address manually on the network.

tzaeschke commented:
> At the moment we can't enforce this, because sometimes the socket will only know it's source address after it has been bound or received the first packet. At that moment it will then want to set the source address on the network, which already has been initialized.
> 
> What I could do is that sockets that got their source address from explicitly binding or receiving the first packet won't write back this source address to the network. Hence, the source address would only need to be set (with connecting to the border router) in the initialization of the network. The drawback there would be that, because sometimes this connecting does not yield the source address sometimes a network won't have a source address set after initialization. We then probably also need the `scion_network_set_addr()` again to be able to set the source address manually on the network.

Okay, I'm not sure what to doe here, just do as you think is best.

koflin commented:
As discussed, we remove the "caching" of the source address in the network. Even though this would have performance benefits when a lot of sockets are created from the same network, managing the source address in the network is probably not worth it for now and makes things more complicated. This removes the need for getters, setters and thread-safety in the network struct.
